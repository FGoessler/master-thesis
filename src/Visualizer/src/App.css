
body {
    margin: 0;
    padding: 0;
    font-family: "Helvetica Neue", Helvetica, Roboto, Arial, sans-serif;
}

.hash {
    font-family: monospace;
}

.highlight {
    font-weight: bold;
}

.full-height {
    height: 100%;
    overflow: scroll;
}

.detailsTable {
    width: 100%;
}
