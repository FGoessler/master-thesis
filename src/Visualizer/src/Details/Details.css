
.details {
    background-color: #f8f5ec;
}

.details td {
    padding: 5px;
}

.details td {
    vertical-align: top;
}

.details ul {
    list-style-type: none;
    padding: 0;
}

.details ul li {
    padding-bottom: 5px;
    margin-bottom: 5px;
    border-bottom: 1px solid lightgrey;
}

.details ul li:last-child {
    border-bottom: none;
}
