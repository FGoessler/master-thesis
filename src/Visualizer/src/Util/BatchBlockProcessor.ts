type Placeholder = null;

// TODO: unit test this!

export default class BatchBlockProcessor<T> {

  private dataFetchingAction: (blockNumber: number) => Promise<T>;
  private dataUpdateCallback?: (datapoints: T[], labels: string[]) => void;

  constructor(
    dataFetchingAction: (blockNumber: number) => Promise<T>,
    dataUpdateCallback?: (datapoints: T[], labels: string[]) => void
  ) {
    this.dataFetchingAction = dataFetchingAction;
    this.dataUpdateCallback = dataUpdateCallback;
  }

  public async startLoading(
    fromBlockNumber: number,
    count: number,
    batchSize: number = 100
  ) {
    let datapoints = [] as T[];
    let labels = [] as string[];
    let nextBlockNumber = fromBlockNumber;

    while (nextBlockNumber > fromBlockNumber - count && nextBlockNumber >= 0) {
      let data = await this.loadBatch(nextBlockNumber, batchSize);
      datapoints = datapoints.concat(data.datapoints);
      labels = labels.concat(data.labels);
      this.sendUpdate(datapoints, labels);
      nextBlockNumber = nextBlockNumber - batchSize;
    }

    return {datapoints, labels};
  }

  private async loadBatch(fromBlockNumber: number, count: number) {
    let datapoints = [] as (T | Placeholder)[];
    let labels = [] as string[];
    let promises = [];
    let nextBlockNumber = fromBlockNumber;
    while (nextBlockNumber > fromBlockNumber - count && nextBlockNumber >= 0) {
      promises.push(this.loadDataAt(nextBlockNumber, datapoints, labels));
      nextBlockNumber = nextBlockNumber - 1;
    }
    await Promise.all(promises);
    return {datapoints: datapoints as T[], labels};
  }

  private async loadDataAt(
    blockNumber: number,
    targetDataArray: (T | Placeholder)[], targetLabelArray: string[]
  ) {
    const targetIndex = targetDataArray.length;

    targetDataArray.push(null);   // push placeholder value
    targetLabelArray.push(`${blockNumber}`);

    targetDataArray[targetIndex] = await this.dataFetchingAction(blockNumber);
  }

  private sendUpdate(datapoints: T[], labels: string[]) {
    if (this.dataUpdateCallback) {
      this.dataUpdateCallback(
        datapoints.slice(0, datapoints.length),
        labels.slice(0, datapoints.length),
      );
    }
  }
}
