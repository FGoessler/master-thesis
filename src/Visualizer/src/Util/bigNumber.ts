const BigNumberLib = require("bignumber.js");

const unitMap = {
  "wei":          "1",
  "kwei":         "1000",
  "ada":          "1000",
  "femtoether":   "1000",
  "mwei":         "1000000",
  "babbage":      "1000000",
  "picoether":    "1000000",
  "gwei":         "1000000000",
  "shannon":      "1000000000",
  "nanoether":    "1000000000",
  "nano":         "1000000000",
  "szabo":        "1000000000000",
  "microether":   "1000000000000",
  "micro":        "1000000000000",
  "finney":       "1000000000000000",
  "milliether":   "1000000000000000",
  "milli":        "1000000000000000",
  "ether":        "1000000000000000000",
  "kether":       "1000000000000000000000",
  "grand":        "1000000000000000000000",
  "einstein":     "1000000000000000000000",
  "mether":       "1000000000000000000000000",
  "gether":       "1000000000000000000000000000",
  "tether":       "1000000000000000000000000000000"
};

export interface BigNumber {
  toNumber(): number;
  toString(base: number): string;
  times(num: number | BigNumber): BigNumber;
  dividedBy(num: number | BigNumber): BigNumber;
}

export class BigNumberHelper {

  static toBigNumber(val: number | string | BigNumber = 0): BigNumber {
    if (val instanceof BigNumberLib) {
      return val as BigNumber;
    }

    if (typeof val === "string" && (val.indexOf("0x") === 0 || val.indexOf("-0x") === 0)) {
      return new BigNumberLib(val.replace("0x", ""), 16);
    }

    return new BigNumberLib((<number> val).toString(), 10);
  }

  static isBigNumber(object: {}) {
    return object instanceof BigNumberLib;
  }

  static toWei(num: number | string | BigNumber, unit: string) {
    return BigNumberHelper.toBigNumber(num).times(BigNumberHelper.getValueOfUnit(unit));
  }

  static fromWei(num: number | string | BigNumber, unit: string) {
    return BigNumberHelper.toBigNumber(num).dividedBy(BigNumberHelper.getValueOfUnit(unit));
  }

  private static getValueOfUnit(unit: string) {
    unit = unit ? unit.toLowerCase() : "ether";
    let unitValue = unitMap[unit];
    if (!unitValue) {
      throw new Error("Unit does not exist. Possible units: " + JSON.stringify(unitMap));
    }
    return new BigNumberLib(unitValue, 10);
  }
}
