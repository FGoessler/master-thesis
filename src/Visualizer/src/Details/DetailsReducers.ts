import { isLoadedSelectedBlock, isLoadingSelectedBlock } from "../Actions/LoadBlock";
import { Block } from "../Util/web3";
import { Action } from "redux";
import { isLoadedAddressChartData, isLoadedAddressInfo, isLoadedContractData } from "../Actions/SelectAddress";
import { BigNumber } from "../Util/bigNumber";
import { ChartData } from "../AddressInspector/BalanceChartCreator";
import { ContractValue } from "../AddressInspector/AddressInfoService";

export interface DetailsState {
  mode: DetailsMode;
  selectedBlock?: Block;
  selectedAddressInfo?: AddressInfo;
}

export enum DetailsMode {
  Nothing,
  Loading,
  BlockDetails,
  AddressDetails
}

export interface AddressInfo {
  address: string;
  balance: BigNumber;
  chartData: ChartData;
  contractData?: ContractValue[];
}

const mergeChartData = (existingData: ChartData, newData: ChartData) => {
  let mergedOldestBlockNr: number;
  if (existingData.oldestBlockNr > 0) {
    mergedOldestBlockNr = Math.min(newData.oldestBlockNr, existingData.oldestBlockNr);
  } else {
    mergedOldestBlockNr = newData.oldestBlockNr;
  }

  return {
    label: newData.label,
    labels: existingData.labels.concat(newData.labels),
    datapoints: existingData.datapoints.concat(newData.datapoints),
    latestBlockNr: Math.max(newData.latestBlockNr, existingData.latestBlockNr),
    oldestBlockNr: mergedOldestBlockNr
  };
};

export const detailsReducer = (state: DetailsState = {
                                 mode: DetailsMode.Nothing,
                                 selectedBlock: undefined,
                                 selectedAddressInfo: undefined
                               },
                               action: Action): DetailsState => {
  if (isLoadedAddressInfo(action)) {
    return {
      mode: DetailsMode.AddressDetails,
      selectedAddressInfo: {
        address: action.address,
        balance: action.balance,
        chartData: {label: "", datapoints: [], labels: [], latestBlockNr: 0, oldestBlockNr: 0},
      }
    };
  } else if (isLoadedSelectedBlock(action)) {
    return {
      mode: DetailsMode.BlockDetails,
      selectedBlock: action.block
    };
  } else if (
    isLoadedAddressChartData(action) &&
    state.mode === DetailsMode.AddressDetails &&
    state.selectedAddressInfo
  ) {
    return {...state,
      selectedAddressInfo: {...state.selectedAddressInfo,
        chartData: mergeChartData(state.selectedAddressInfo.chartData, action.chartData)
      }
    };
  } else if (
    isLoadedContractData(action) &&
    state.mode === DetailsMode.AddressDetails &&
    state.selectedAddressInfo
  ) {
    return {...state,
      selectedAddressInfo: {...state.selectedAddressInfo,
        contractData: action.contractData
      }
    };
  } else if (isLoadingSelectedBlock(action)) {
    return {
      mode: DetailsMode.Loading,
      selectedBlock: undefined,
      selectedAddressInfo: undefined
    };
  } else {
    return state;
  }
};
