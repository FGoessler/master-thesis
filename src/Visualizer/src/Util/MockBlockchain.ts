
import { Block, Eth, Web3, Transaction, TransactionReceipt } from "./web3";
import { BigNumber, BigNumberHelper } from "./bigNumber";

export class MockBlock implements Block {
  number: number = 1;
  hash: string = "0x2a";
  parentHash: string = "0x1a";
  timestamp: number = 100000;
  transactions: (string | Transaction)[] = [];

  constructor(num: number = 1) {
    this.number = num;
  }
}

export class MockTransaction implements Transaction {
  hash: string = "0xffff1a";
  from: string = "0xdd1a";
  to: string  = "0xdd2b";
  value: BigNumber = BigNumberHelper.toBigNumber(0);
  input: string = "";
  transactionReceipt?: TransactionReceipt;
  isContractDeployment: boolean = false;
}

export default class MockWeb3 implements Web3 {

  public eth: Eth;

  constructor() {
    this.eth = new MockEth();
  }
}

export class MockEth implements Eth {

  public transactionReceipts: { [key: string]: TransactionReceipt };
  public blocks: Block[];
  public balances: {};

  constructor() {
    this.transactionReceipts = {
      "0xffff1a": {
        blockHash: "0x2a",
        blockNumber: 1,
        transactionHash: "0xffff1a",
        gasUsed: 100000,
        logs: []
      }
    };

    this.blocks = [
      {
        number: 0,
        hash: "0x1a",
        parentHash: "",
        timestamp: 100000,
        transactions: []
      },
      {
        number: 1,
        hash: "0x2a",
        parentHash: "0x1a",
        timestamp: 100000,
        transactions: [
          new MockTransaction()
        ]
      },
      {
        number: 2,
        hash: "0x3a",
        parentHash: "0x2a",
        timestamp: 100000,
        transactions: []
      }
    ];

    this.balances = {
      "0x1234": BigNumberHelper.toBigNumber("9900000")
    };
  }

  contract(abi: any[]) {
    return {};
  }

  getBlock(hash: number | string, inclTransactions: boolean, cb: (err?: Error, b?: Block) => void): void {
    if (typeof hash === "number") {
      cb(undefined, this.blocks[hash]);
    } else {
      if (hash === "latest") {
        cb(undefined, this.blocks[this.blocks.length - 1]);
      } else {
        const foundBlock = this.blocks.find((b) => { return b.hash === hash; });
        if (foundBlock) {
          cb(undefined, foundBlock);
        } else {
          cb(new Error(`Block ${hash} not found!`), undefined);
        }
      }
    }
  }

  getTransactionReceipt(hash: string, cb: (err?: Error, b?: TransactionReceipt) => void): void {
    if (this.transactionReceipts[hash]) {
      cb(undefined, this.transactionReceipts[hash]);
    } else {
      cb(new Error(`Transaction ${hash} not found!`), undefined);
    }
  }

  getBalance(address: string, cb: ((err?: Error, b?: BigNumber) => void)): void;
  getBalance(address: string, block: number | string, cb: (err?: Error, b?: BigNumber) => void): void;
  getBalance(address: string,
             blockOrCb: number | string | ((err?: Error, b?: BigNumber) => void),
             cb?: ((err?: Error, b?: BigNumber) => void)): void {
    let callback: (err?: Error, b?: BigNumber) => void;
    if (typeof blockOrCb === "function") {
      callback = blockOrCb;
    } else if (cb && typeof cb === "function") {
      callback = cb;
    } else {
      throw new Error("Missing callback!");
    }

    const balance = this.balances[address];
    if (balance) {
      callback(undefined, balance);
    } else {
      callback(new Error(`Address ${address} not found!`), undefined);
    }
  }
}
