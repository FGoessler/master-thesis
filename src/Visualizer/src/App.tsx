import * as React from "react";
import "./App.css";
import List from "./List/List";
import Details from "./Details/Details";
import { ListState } from "./List/ListReducers";
import { DetailsState } from "./Details/DetailsReducers";

export interface AppState {
  details: DetailsState;
  list: ListState;
}

export default class App extends React.Component<{}, {}> {

  render() {
    return (
      <div className="container-fluid" style={{height: "100vh"}}>
        <div className="row full-height">
          <div className="col-xs-4 list full-height">
            <List />
          </div>
          <div className="col-xs-8 details full-height">
            <Details />
          </div>
        </div>
      </div>
    );
  }
}
