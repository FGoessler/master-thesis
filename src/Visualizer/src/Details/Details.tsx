import * as React from "react";
import BlockDetails from "../BlockInspector/BlockDetails";
import { connect } from "react-redux";
import { Block } from "../Util/web3";
import "./Details.css";
import { AppState } from "../App";
import { AddressInfo, DetailsMode } from "./DetailsReducers";
import AddressDetails from "../AddressInspector/AddressDetails";
import { loadAddressChartData, selectAddress } from "../Actions/SelectAddress";
import { Dispatch } from "redux";

interface DetailsProps {
  mode: DetailsMode;
  selectedBlock?: Block;
  selectedAddressInfo?: AddressInfo;
  addressSelectedHandler: (address: string) => void;
  onRequestOlderBlocks: (address: string, nextBlock: number) => void;
}

export class Details extends React.Component<DetailsProps, {}> {

  render() {
    switch (this.props.mode) {
      case DetailsMode.Nothing:
        return (<h3>Nothing selected.</h3>);

      case DetailsMode.Loading:
        return (<h3>Loading...</h3>);

      case DetailsMode.BlockDetails:
        if (!this.props.selectedBlock) {
          throw new Error("DetailsMode.BlockDetails but selectedBlock is missing.");
        }
        return (
          <BlockDetails
            block={this.props.selectedBlock}
            addressSelectedHandler={this.props.addressSelectedHandler}
          />
        );

      case DetailsMode.AddressDetails:
        if (!this.props.selectedAddressInfo) {
          throw new Error("DetailsMode.AddressDetails but selectedAddressInfo is missing.");
        }
        return (
          <AddressDetails
            address={this.props.selectedAddressInfo.address}
            currentBalance={this.props.selectedAddressInfo.balance}
            chartData={this.props.selectedAddressInfo.chartData}
            contractData={this.props.selectedAddressInfo.contractData}
            onRequestOlderBlocks={this.props.onRequestOlderBlocks}
          />
        );

      default:
        throw new Error(`Missing case - no details defined for '${this.props.mode}'.`);
    }
  }
}

const mapStateToProps = (state: AppState): Partial<DetailsProps> => {
  return state.details;
};

const mapDispatchToProps = (dispatch: Dispatch<AppState>): Partial<DetailsProps> => {
  return {
    addressSelectedHandler: (address: string) => {
      dispatch(selectAddress(address));
    },
    onRequestOlderBlocks: (address: string, nextBlock: number) => {
      dispatch(loadAddressChartData(address, nextBlock));
    }
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Details);
