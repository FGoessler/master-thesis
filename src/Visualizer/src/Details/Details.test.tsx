import * as React from "react";
import * as ReactDOM from "react-dom";
import configureMockStore from "redux-mock-store";
import { Provider } from "react-redux";
import { MockBlock } from "../Util/MockBlockchain";
import Details from "./Details";
import { DetailsMode } from "./DetailsReducers";
import { BigNumberHelper } from "../Util/bigNumber";
import { AppState } from "../App";

it("renders block details", () => {
  const mockStore = configureMockStore()({details: {
    mode: DetailsMode.BlockDetails,
    selectedBlock: new MockBlock(1)
  }});
  mockStore.dispatch = jest.fn();

  const div = document.createElement("div");
  ReactDOM.render(<Provider store={mockStore}><Details/></Provider>, div);
  expect(div).toMatchSnapshot();
});

it("renders address details", () => {
  const initialState: Partial<AppState> = {details: {
    mode: DetailsMode.AddressDetails,
    selectedAddressInfo: {
      address: "0x1234",
      balance: BigNumberHelper.toBigNumber(5000),
      chartData: {
        label: "",
        datapoints: [],
        labels: [],
        latestBlockNr: 0,
        oldestBlockNr: 0
      }
    }
  }};
  const mockStore = configureMockStore()(initialState);
  mockStore.dispatch = jest.fn();

  const div = document.createElement("div");
  ReactDOM.render(<Provider store={mockStore}><Details/></Provider>, div);
  expect(div).toMatchSnapshot();
});

it("renders nothing selected state", () => {
  const mockStore = configureMockStore()({details: {
    mode: DetailsMode.Nothing
  }});
  mockStore.dispatch = jest.fn();

  const div = document.createElement("div");
  ReactDOM.render(<Provider store={mockStore}><Details/></Provider>, div);
  expect(div).toMatchSnapshot();
});

it("renders loading state", () => {
  const mockStore = configureMockStore()({details: {
    mode: DetailsMode.Loading
  }});
  mockStore.dispatch = jest.fn();

  const div = document.createElement("div");
  ReactDOM.render(<Provider store={mockStore}><Details/></Provider>, div);
  expect(div).toMatchSnapshot();
});
