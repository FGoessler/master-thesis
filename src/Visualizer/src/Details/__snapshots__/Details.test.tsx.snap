// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders address details 1`] = `
<div>
  <div
    data-reactroot=""
  >
    <h3>
      0x1234
    </h3>
    <table>
      <tbody>
        <tr>
          <td>
            Address
          </td>
          <td
            class="hash"
          >
            0x1234
          </td>
        </tr>
        <tr>
          <td>
            Balance
          </td>
          <td>
            5000
          </td>
        </tr>
      </tbody>
    </table>
    <div>
      No Contract Data
    </div>
  </div>
</div>
`;

exports[`renders block details 1`] = `
<div>
  <div
    data-reactroot=""
  >
    <h3>
      0x2a
    </h3>
    <table
      class="detailsTable"
    >
      <tbody>
        <tr>
          <td>
            Hash
          </td>
          <td
            class="hash"
          >
            0x2a
          </td>
        </tr>
        <tr>
          <td>
            Number
          </td>
          <td>
            1
          </td>
        </tr>
        <tr>
          <td>
            Timestamp
          </td>
          <td>
            Fri, 02 Jan 1970 03:46:40 GMT
          </td>
        </tr>
        <tr>
          <td>
            Parent Hash
          </td>
          <td
            class="hash"
          >
            0x1a
          </td>
        </tr>
        <tr>
          <td>
            Transactions
          </td>
          <td>
            <ul />
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`renders loading state 1`] = `
<div>
  <h3
    data-reactroot=""
  >
    Loading...
  </h3>
</div>
`;

exports[`renders nothing selected state 1`] = `
<div>
  <h3
    data-reactroot=""
  >
    Nothing selected.
  </h3>
</div>
`;
